lib/src/functions/map.dart

```diff
@@ -131,36 +152,41 @@ final _hasKey = _function("has-key", r"$map, $key, $keys...", (arguments) {
   return SassBoolean(map.contents.containsKey(keys.last));
 });
 
-/// Updates the specified value in [map] by applying the [modify()] callback to
+/// Updates the specified value in [map] by applying the [modify] callback to
 /// it, then returns the resulting map.
 ///
 /// If more than one key is provided, this means the map targeted for update is
 /// nested within [map]. The multiple [keys] form a path of nested maps that
 /// leads to the targeted map. If any value along the path is not a map, and
-/// [overwrite] is true, this inserts a new map at that key and overwrites the
-/// current value. Otherwise, this fails and returns [map] with no changes.
-SassMap _modify(SassMap map, List<Value> keys, Value modify(Value old),
-    [bool overwrite = true]) {
-  SassMap _modifyNestedMap(SassMap map, int index) {
+/// `modify(null)` returns null, this inserts a new map at that key and
+/// overwrites the current value. Otherwise, this fails and returns [map] with
+/// no changes.
+///
+/// If no keys are provided, this passes [map] directly to modify and returns
+/// the result.
+Value _modify(SassMap map, Iterable<Value> keys, Value modify(Value old)) {
+  var keyIterator = keys.iterator;
+  SassMap _modifyNestedMap(SassMap map, [Value newValue]) {
     var mutableMap = Map.of(map.contents);
-    var key = keys[index];
+    var key = keyIterator.current;
 
-    if (index == keys.length - 1) {
-      mutableMap[key] = modify(mutableMap[key]);
+    if (!keyIterator.moveNext()) {
+      mutableMap[key] = newValue ?? modify(mutableMap[key]);
       return SassMap(mutableMap);
     }
 
     var nestedMap = mutableMap[key]?.tryMap();
-    if (nestedMap == null && !overwrite) {
-      return SassMap(mutableMap);
+    if (nestedMap == null) {
+      newValue = modify(null);
+      if (newValue == null) return SassMap(mutableMap);
     }
 
     nestedMap ??= const SassMap.empty();
-    mutableMap[key] = _modifyNestedMap(nestedMap, index + 1);
+    mutableMap[key] = _modifyNestedMap(nestedMap, newValue);
     return SassMap(mutableMap);
   }
 
-  return _modifyNestedMap(map, 0);
+  return keyIterator.moveNext() ? _modifyNestedMap(map) : modify(map);
```

Comment:
Could this instead be the base case of `_modify()`?

Reply:
Do you mean making this like `if (keys.isEmpty) return modify(map)` at the top of `_modify()` or as a base case of `_modifyNestedMap()`?

For the latter, since it deals with mutating a key within a map, it doesn't really work when there are no keys at all. For the former, `keyIterator.moveNext()` needs to be called once before `_modifyNestedMap()` (to ensure that `keyIterator.current` exists), so making the check for empty keys separate doesn't really make things any simpler.